### ISSUE TITLE
canonicalizeEmail removes spaces inside emails

### ISSUE THREAD

**Issue by riz0id** [Porges/email-validate-hs#66]
I ran into a issue with the [`canonicalizeEmail` function](https://hackage.haskell.org/package/email-validate-2.3.2.18/docs/Text-Email-Validate.html#v:canonicalizeEmail) where whitespace interspersed between atoms of an email were trimmed from the given email, for example:

```haskell
>>> :set -XOverloadedStrings
>>> import Data.ByteString (ByteString)
>>> import Text.Email.Validate (canonicalizeEmail)
>>> canonicalizeEmail ("alice. cool. email@foo.bar" :: ByteString)
(Just "alice.cool.email@foo.bar") :: Maybe ByteString
```

This was not originally the case for  [`canonicalizeEmail` function](https://hackage.haskell.org/package/email-validate-2.3.2.18/docs/Text-Email-Validate.html#v:canonicalizeEmail). Emails containing whitespace were only accepted after #3 was raised. I was not able to find where RFC 5322 listed whitespace as valid characters for the email grammar however. Shouldn't the correct implementation reject emails containing whitespace? i.e.

 ```haskell
>>> canonicalizeEmail ("alice. cool. email@foo.bar" :: ByteString)
Nothing :: Maybe ByteString
```

I wasn't able to find where RFC 5322 allowed for whitespace outside quotations in the grammar it specified for email address. However, even if `"alice. cool. email@foo.bar"` constitutes a valid email address according to RFC 5322, stripping whitespace within emails means that the emails `"alice. cool. email@foo.bar"` and `"alice.cool.email@foo.bar"` are essentially equivalent according to [`canonicalizeEmail` function](https://hackage.haskell.org/package/email-validate-2.3.2.18/docs/Text-Email-Validate.html#v:canonicalizeEmail):

```haskell
>>> import Data.Function (on)
>>> ((==) `on` canonicalizeEmail) "alice. cool. email@foo.bar" "alice.cool.email@foo.bar"
True
```

Of course I think trimming surrounding whitespace is probably reasonable since it does not change the email address in any meaningful way. I do believe that the string `"alice. cool. email@foo.bar"` is meaningfully different from the string `"alice.cool.email@foo.bar"`, so I don't believe [`canonicalizeEmail` function](https://hackage.haskell.org/package/email-validate-2.3.2.18/docs/Text-Email-Validate.html#v:canonicalizeEmail) should behave as if they aren't.

Ideally I would like [`canonicalizeEmail`](https://hackage.haskell.org/package/email-validate-2.3.2.18/docs/Text-Email-Validate.html#v:canonicalizeEmail) to outright reject emails containing whitespace. If that isn't desirable for whatever reason, then I would be really nice for [`canonicalizeEmail`](https://hackage.haskell.org/package/email-validate-2.3.2.18/docs/Text-Email-Validate.html#v:canonicalizeEmail) to not strip whitespace in emails. Especially since the original issue that motivated the change which allowed [`canonicalizeEmail`](https://hackage.haskell.org/package/email-validate-2.3.2.18/docs/Text-Email-Validate.html#v:canonicalizeEmail) to accept emails containing spaces was made to meet the needs of an individual at the cost of making  [`canonicalizeEmail`](https://hackage.haskell.org/package/email-validate-2.3.2.18/docs/Text-Email-Validate.html#v:canonicalizeEmail) diverge from the RFC that the rest of the library adheres to.

I'm happy to make either of the changes I proposed here if you agree with me, just let me know what I can do.

**Comment by Porges** [comment]
Hi @riz0id, the whitespace in the grammer is the "CFWS" production. It is optionally allowed in `local-part` (the left hand side of the email) via:

```
   local-part      =   dot-atom / quoted-string / obs-local-part
   dot-atom        =   [CFWS] dot-atom-text [CFWS]
   dot-atom-text   =   1*atext *("." 1*atext)
   atom            =   [CFWS] 1*atext [CFWS]
```

The removal of CFWS when canonicalizing is due to the paragraph [after these productions](https://www.rfc-editor.org/rfc/rfc5322#section-3.2.3) (my **highlighting**):

>    Both atom and dot-atom are interpreted as a single unit, comprising the string of characters that make it up. **Semantically, the optional comments and FWS surrounding the rest of the characters are not part of the atom;** the atom is only the run of atext characters in an atom, or the atext and "." characters in a dot-atom.

**Comment by Porges** [comment]
Note that removing spaces & comments from the email is basically all that `canonicalizeEmail` does. So if you don't want this, you can instead use the `isValid` function and keep around the original `ByteString`.

**Comment by riz0id** [comment]
Okay thanks.
